Swift/BRCrypto/BRCryptoWallet.swift

```diff
@@ -169,18 +169,252 @@ public final class Wallet: Equatable {
+    /// reject.  In those casaes the minimum amount is above zero.
+    ///
+    /// In cases where `balance < amount + fee` then .insufficientFunds is returned.  The
+    /// .insufficientFunds check respects the wallet from which fees are extracted.
+    ///
+    /// This is an synchronous function that returns immediately but will call `completion` once
+    /// the maximum has been determined.
+    ///
+    /// The returned Amount is always in the wallet's currencyh.
+    ///
+    /// - Parameters:
+    ///   - target: the target address
+    ///   - fee: the network fees
+    ///   - completion: the handler for the results
+    ///
+    public func estimateLimitMinimum (target: Address,
+                                      fee: NetworkFee,
+                                      completion: @escaping Wallet.EstimateLimitHandler) {
+        estimateLimit (asMaximum: false, target: target, fee: fee, completion: completion)
+    }
+
+    ///
+    /// Internal function to handle limit estimation
+    ///
+    internal func estimateLimit (asMaximum: Bool,
+                                 target: Address,
+                                 fee: NetworkFee,
+                                 completion: @escaping Wallet.EstimateLimitHandler) {
+        var needFeeEstimate: BRCryptoBoolean = CRYPTO_TRUE
+        var isZeroIfInsuffientFunds: BRCryptoBoolean = CRYPTO_FALSE;
+
+        // This `amount` is in the `unit` of `wallet`
+        guard let amount = cryptoWalletEstimateLimit (self.core,
+                                                      (asMaximum ? CRYPTO_TRUE : CRYPTO_FALSE),
+                                                      target.core,
+                                                      fee.core,
+                                                      &needFeeEstimate,
+                                                      &isZeroIfInsuffientFunds)
+            .map ({ Amount (core: $0, take: false)})
+            else {
+                // This is extraneous as `cryptoWalletEstimateLimit()` always returns an amount
+                completion (Result.failure (LimitEstimationError.insufficientFunds))
+                return;
+        }
+
+        // If we don't need an estimate, then we invoke `completion` and skip out immediately.  But
+        // include a check on a zero amount - which indicates insufficient funds.
+        if CRYPTO_FALSE == needFeeEstimate {
+            system.queue.async {
+                completion (CRYPTO_TRUE == isZeroIfInsuffientFunds && amount.isZero
+                    ? Result.failure (LimitEstimationError.insufficientFunds)
+                    : Result.success (amount))
+            }
+            return
+        }
+
+        // We need an estimate of the fees.
+
+        // The currency for the fee
+        let currencyForFee = fee.pricePerCostFactor.currency
+
+        guard let walletForFee = self.manager.wallets
```

Comment:
This sort of speaks to this function hanging off the manager, no?

Reply:
I'd like to move to all 'active' functions to be at WalletManager or System level.  Maybe this is the place to start.

Reply:
Will defer.